l2rpn_baselines/utils/BaseDeepQ.py

```diff
@@ -114,32 +114,53 @@ def predict_movement(self, data, epsilon, batch_size=None):
         if batch_size is None:
             batch_size = data.shape[0]
 
-        rand_val = np.random.random(batch_size)
-        q_actions = self._model.predict(data, batch_size=batch_size)
-
-        opt_policy = np.argmax(np.abs(q_actions), axis=-1)
-        opt_policy[rand_val < epsilon] = np.random.randint(0, self._action_size, size=(np.sum(rand_val < epsilon)))
-        return opt_policy, q_actions[0, opt_policy]
+        q_actions = self._model.predict(data, batch_size=batch_size)  # q value of each action
+        opt_policy = np.argmax(q_actions, axis=-1)
+        if epsilon > 0.:
+            rand_val = np.random.random(batch_size)
+            opt_policy[rand_val < epsilon] = np.random.randint(0, self._action_size, size=(np.sum(rand_val < epsilon)))
+        return opt_policy, q_actions[np.arange(batch_size), opt_policy], q_actions
 
     def train(self, s_batch, a_batch, r_batch, d_batch, s2_batch, tf_writer=None, batch_size=None):
-        """Trains network to fit given parameters"""
+        """Trains network to fit given parameters:
+        Parameters
+        ----------
+        s_batch:
+            the state vector (before the action is taken)
+        a_batch:
+            the action taken
+        s2_batch:
+            the state vector (after the action is taken)
+        d_batch:
+            says whether or not the episode was over
+        r_batch:
+            the reward obtained this step
+
+        see https://towardsdatascience.com/dueling-double-deep-q-learning-using-tensorflow-2-x-7bbbcec06a2a
+        for the update rules
+        """
         if batch_size is None:
             batch_size = s_batch.shape[0]
 
         # Save the graph just the first time
         if tf_writer is not None:
             tf.summary.trace_on()
-        targets = self._model.predict(s_batch, batch_size=batch_size)
+        target = self._model.predict(s_batch, batch_size=batch_size)
```

Comment:
Same here for calls to `predict`, can be a LOT faster by using the `__call__` interface.